core/src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathBasedFilter.java

```diff
@@ -77,31 +82,39 @@ private TokenFilter evaluate(String name, FilterPath[] filters) {
             }
 
             if ((nextFilters != null) && (nextFilters.isEmpty() == false)) {
-                return new FilterPathBasedFilter(nextFilters.toArray(new FilterPath[nextFilters.size()]));
+                return new FilterPathBasedFilter(inclusive, nextFilters.toArray(new FilterPath[nextFilters.size()]));
             }
         }
         return NO_MATCHING;
     }
 
+
     @Override
     public TokenFilter includeProperty(String name) {
-        TokenFilter include = evaluate(name, filters);
-        if (include == MATCHING) {
-            return TokenFilter.INCLUDE_ALL;
-        }
-        if (include == NO_MATCHING) {
-            return null;
+        TokenFilter filter = evaluate(name, filters);
+        if (inclusive) {
+            if (filter == MATCHING) {
+                return TokenFilter.INCLUDE_ALL;
+            } else if (filter == NO_MATCHING) {
+                return null;
+            }
+        } else {
+            if (filter == MATCHING) {
+                return null;
+            } else if (filter == NO_MATCHING) {
+                return TokenFilter.INCLUDE_ALL;
+            }
         }
```

Comment:
Yeah, I'll agree this time making it more compact makes it easier to see the pattern. I'll change it.